Large internal changes

realhidden opened the pull request:
First of all, this is really nice library. I tried to use it to resolve 10k dns records in a meaningful way, and the retry / timeout logic resulted in mostly timeouts. The memory usage and gc cpu usage also went up if I did 40k queries parallel.

So I started to dig around, and changed the _id and _queries structure to a more static one, changing the timeout and retry logic, extending the host parameter to accept arrays and randomly pick from them. 

There was an issue #15 which also needed to be resolved for my use-case, so I extended the library to support them.

The only "breaking" change is that the host parameter is not an optional one any more, that's why I modified the tests - otherwise for an average user it looks and works the same.

This PR has a slight memory advantage over the original one, the host and retry logic might be useful for someone else as well.

Let me know if you want to have any modifications, etc.

realhidden committed:
up the dependency + .idea to gitignore

realhidden committed:
Large internal rewrite

+ maxQueries option
+ maxRedirects option
+ timeoutChecks option

mafintosh commented:
Interesting!

Did this allow you to do the 40k queries in parallel?
Can you put some words here on how you changed the retry and timeout stuff?

realhidden committed:
Make timeout logic work with setRetries

realhidden commented:
Yep, 40k query is ok, but it requires some time. I used 4 open dns servers (google and cloudflare ones), timeout set to 1 sec, retry to 30, maxRedirects to 5, maxQueries to 50k. If the average response time is around 15ms, it can resolve 16k records in 1 minute, yet the domain list I used has some nxdomains, so it finished in 30sec. Not bad at all.

The timeout / retry logic now a simple setInterval which tells us if we need to retry it based on when we sent the first packet (Date.now()) and how many retries we should have done so far. If enough time has passed, we can send a new packet to some server and hope for the best.

I modified the source so the setRetries should work now as expected, since we don't use it, I didn't notice that it broke.

realhidden committed:
Remove a semicolon

silverwind commented on README.md:
```diff
@@ -32,6 +32,9 @@ Create a new DNS socket instance. The `options` object includes:
 - `socket` *Object*: A custom dgram socket. Default: A `'udp4'` socket.
 - `timeout` *Number*: Total timeout in milliseconds after which a `'timeout'` event is emitted. Default: 7500.
+- `maxQueries` *Number*: Each request has an id, this is stored as static sized array. maxQueries is the size of this array, limiting the max number of inflight requests. Default: 10000.
+- `maxRedirects` *Number*: If you query for a single `A` record and get back `CNAME`, the lib will try to follow the chain and resolve the `CNAME` to A. The maximum number of steps is defined by the `maxRedirects`. Default: 0
```
I think default should be to follow redirects, maybe with a default maximum value of 50.

realhidden replied:
I didn't want to break existing functionality of the nofollow. 50 is kind of extensive I think, CNAME points to CNAME points to CNAME .... 50 times, more realistic is just a few steps, like CNAME to CNAME to A record.

silverwind replied:
On second thought, I'm ok with having it off by default, to avoid the breaking change.

silverwind commented on index.js:
```diff
@@ -80,37 +87,85 @@ DNS.prototype.bind = function (...args) {
-    q.tries.shift()
-    this.socket.send(q.buffer, 0, q.buffer.length, q.port, q.host)
+    q.tries++
+    this.socket.send(q.buffer, 0, q.buffer.length, q.port, typeof q.host === 'object' ? q.host[Math.floor(q.host.length * Math.random())] : q.host)
```
use `Array.isArray(q.host)` to detect arrays.

silverwind commented on index.js:
```diff
@@ -80,37 +87,85 @@ DNS.prototype.bind = function (...args) {
+
+  const id = this._getNextEmptyId()
+  if (id === -1) {
+    q.callback('Query array is full!')
```
Is this an error? If so, wrap the string with `new Error()`.

silverwind commented on index.js:
```diff
@@ -124,27 +179,21 @@ DNS.prototype._onmessage = function (buffer, rinfo) {
   if (message.type === 'response' && message.id) {
-    const i = this._ids.indexOf(message.id)
-    const q = i > -1 ? this._queries[i] : null
+    // const i = this._ids.indexOf(message.id);
```
No commented out code please.

silverwind commented:
> The only "breaking" change is that the host parameter is not an optional one any more, that's why I modified the tests - otherwise for an average user it looks and works the same.

Node.js would default to `127.0.0.1`, are you sure it's necessary to make `host` required?

realhidden committed:
Code review changes

-Array.isArray
- host default parameter
- missing Error
- remove commented code

realhidden committed:
Mark host parameter as optional in the documentation

realhidden commented:
Changed the code to fit the review, host default parameter was reintroduced. The CNAME auto redirect / resolve function ca be a breaking change, so it might need a new mayor semver or we can leave it to 0 as default.

silverwind commented:
I have a use case where I do 1k parallel queries against localhost, will try to get a performance comparison on that one.

silverwind commented:
Some results against local caching recursor (same hostname):

```` bash
# 100k queries, 200 concurrently
master:  0.188ops/sec  5309.5606msec/op      +-57.3209msec/op (1.08%)
pr16:    0.191ops/sec  5239.5435msec/op      +-29.0529msec/op (0.55%)

# 100k queries, 1k concurrently
master:  0.188ops/sec  5307.016msec/op       +-48.5663msec/op (0.92%)
pr16:    0.192ops/sec  5218.2297msec/op      +-25.5149msec/op (0.49%)
````
There seems to be a very minor performance gain. Most time is probably lost on the network.
